Use Ember.Object.setProperties to update props

dustinfarris opened the pull request:
This is a different approach to the concerns raised in #30

### Using `Ember.Object.setProperties`

This greatly simplifies the code and transfers optimization responsibility to Ember, where it should stay if at all possible.

The properties work like this:

On init (and after every state change via redux.subscribe), the user props are computed using the user-supplied `mapStateToComputed` function.  The resulting object might look like e.g.:

```js
// props
{
  users: [ { id: 1, name: 'Bob' } ]
}
```

The resulting props object is set on a private `_props` property on the component via `setProperties`, e.g.:

```js
this.get('_props').setProperties(props);
```

"read-only" aliases are set up to match what the user expects

```js
// Simplified
Ember.defineProperty(this, 'user', Ember.computed.readOnly('_props.user'));
```

This allows us to update the "private" component properties using redux.subscribe, and the user continues to consume the public read-only props (which are just aliases).  It is important to emphasize that _no additional state is being created_—just a pathway that allows us to make updates to the component and still keep everything locked down.

----

cc @brettburley

toranb commented:
@dustinfarris I like the simplicity of this for sure. One blocker I see currently is that setProperties is called with a new object each time (w/ Object.assign) likely resulting in a notifyPropertyChange for each property in the component. I can't say this is the case for certain so I'll need you and I to do some home work here.

Here is what I would like to accompany this PR regarding test coverage that I couldn't seem to find myself searching around for a few minutes tonight.

1. Throw out my monkey patch for `updateProps` and instead look deeper at what is truly invoked at the ember metal layer when notifyPropertyChange is fired.
2. Use this to get a new snapshot/metric before the changes you've set out here
3. Confirm your changes (and setProperties) doesn't incur any additional work from ember (as I suspect it does looking at setProperties here)

@brettburley if you have any working knowledge of ember internals (ie: what gets notified /when) it would greatly improve our ability to confirm we aren't back in over-notify land. I want to make sure we go above and beyond if we plan to refactor this because @justinpark opened an issue related to this very problem in #37 recently.

brettburley commented on addon/components/connect.js:
```diff
@@ -14,13 +14,38 @@ var connect = function(mapStateToComputed, mapDispatchToActions) {
+        var transformPropsToPrivate = function(props) {
+            // Transform object to new object with private keys
+            return Object.keys(props).reduce(function(privateState, key) {
+                return Object.assign(privateState, {
```
Any advantage to using `Object.assign` here, rather than just assigning `privateState[getPrivateKey(key)] = props[key]`? I just ran a quick test in the console comparing the two and saw a ~50x difference in speed. Granted, I'm not sure how critical a few ms saved here is in the grand scheme of things.

dustinfarris replied:
I don't think there's an advantage to `Object.assign`.  I'll make the change.

brettburley commented on addon/components/connect.js:
```diff
@@ -14,13 +14,38 @@ var connect = function(mapStateToComputed, mapDispatchToActions) {
+         * Transform a key into its private equivalent by prepending an underscore
+         */
+        var getPrivateKey = function(key) {
+            return `_${key}`;
```
Any thoughts on storing the keys on the component itself vs. a private Ember.Object property like `_props` (ala ember-data's `_data` property)? In the latter, no key transformation is necessary which seems like a plus. Plus all the data is stored together (and eliminates issues where component users try to use an underscored version of a property for their own purposes and might conflict).

dustinfarris replied:
Yeah, I like that!  I'll adjust and we can see how it looks.

brettburley commented:
> One blocker I see currently is that setProperties is called with a new object each time (w/ Object.assign) likely resulting in a notifyPropertyChange for each property in the component.

Fortunately, `setProperties` will noop when the new value is equal to the current value. So it's safe to call it with an entirely new hash. We will see the issue we discovered here https://github.com/toranb/ember-redux/pull/30#issuecomment-258061853 where unchanged objects may be re-set as a result of the original being mutated by Ember at runtime. But the notification frequency should match the behavior with the current check here: https://github.com/toranb/ember-redux/blob/master/addon/components/connect.js#L52

Regarding tests, I think we should definitely be counting number of re-renders. One way to do that is like I did here by tracking `didRender`s: https://github.com/toranb/ember-redux/pull/30, but there might be other ways to count the same. It would probably be nice to add a few tests that only change properties that aren't rendered and verifying that the component isn't re-rendered.

We could also add some tests that set up observers on the properties and ensure the observers don't fire when the property doesn't change.

toranb commented:
@dustinfarris you mind implementing the re-render checks that @brettburley mentioned above? I was reading up on the [mixin](https://github.com/emberjs/ember.js/blob/v2.9.1/packages/ember-runtime/lib/mixins/observable.js#L30) that does observer work but I think re-render maybe the most appropriate give this approach. The biggest challenge I have with re-render only is that if I have 2 properties on the component and only 1 changes - I assume a re-render is invoked but did we (or did we not) notifyPropertyChange for both props?

toranb commented:
@dustinfarris could you push up the object.assign changes @brettburley mentioned? Without them we do indeed see a regression related to the notifyPropertyChange (using my new test in PR #42 )

here is the diff after using setProperties as you have it above

![screen shot 2016-11-11 at 8 13 09 pm](https://cloud.githubusercontent.com/assets/147411/20234971/95bc8db2-a84b-11e6-9c66-00ce50c7145e.png)

brettburley commented:
I don't think `Object.assign` will change the behavior at all. Let's see after #42 is updated if the counts still mismatch... I would expect them to be the same in this case.

toranb commented:
@dustinfarris could you rebase this now that PR #42 have merged in? I found that your work w/ setProperties here is now regression free because ember already do the heavy lifting before it decides to notifyPropertyChange

![screen shot 2016-11-19 at 3 19 25 pm](https://cloud.githubusercontent.com/assets/147411/20458497/ae918cda-ae6c-11e6-8ef2-e8e61e5d3167.png)

dustinfarris commented:
Rebased

dustinfarris commented:
I am addressing Brett's comments now

dustinfarris commented:
Ok, rebased and updated with @brettburley's suggestions—which simplified the code greatly as it is no longer necessary to transform the props object at all.

I've updated the description to match.  This should be good to go.

toranb commented:
@dustinfarris this is awesome! Just look at all the code we can delete! Thanks for all the hard work here

@brettburley this seems rock solid and ready to merge from my view. You like?

brettburley commented:
This looks great! Thanks for all the work @dustinfarris.

toranb commented:
@dustinfarris @brettburley this is now up on npm as part of the v1.9.0 release
